File: solutions/python/string_to_integer_atoi.py
# Runtime 51 ms Beats 18.98%
# Memory 16.3 MB Beats 22.84%

class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.strip()

        if not s:
            return 0

        sign = 1

        if s[0:1] == "-":
            sign = -1
            s = s[1:]
        elif s[0:1] == "+":
            sign = 1
            s = s[1:]

        sArray = list(s)

        if len(sArray) == 0:
            return 0

        result = ""
        for c in sArray:
            if c.isnumeric():
                result += c
            else:
                break

        if result == "":
            return 0

        result = int(result) * sign
        upLimit = (2**31) - 1
        bottomLimit = -2**31
        result = upLimit if result > upLimit else bottomLimit if result < bottomLimit else result

        return int(result)

File: solutions/csharp/StringToIntegerAtoi.cs
// Runtime 70 ms Beats 62.30%
// Memory 41 MB Beats 10.97%

namespace StringtoIntegerAtoi {
    public class Solution {
        public int MyAtoi(string s) {
            s = s.Trim();

            if (string.IsNullOrEmpty(s)) {
                return 0;
            }

            int sign = 1;

            if (s[0] == '-') {
                sign = -1;
                s = s.Substring(1);
            } else if (s[0] == '+') {
                sign = 1;
                s = s.Substring(1);
            }

            char[] sArray = s.ToCharArray();

            if (sArray.Length == 0) {
                return 0;
            }

            string result = "";

            foreach (char c in sArray) {
                if (char.IsNumber(c)) {
                    result += c;
                } else {
                    break;
                }
            }

            if (string.IsNullOrEmpty(result)) {
                return 0;
            }

            int upLimit = (int)Math.Pow(2, 31) - 1;
            int bottomLimit = -(int)Math.Pow(2, 31);
            int parsedResult;
            try {
                parsedResult = checked(int.Parse(result)) * sign;
            } catch (OverflowException) {
                parsedResult = sign == 1 ? upLimit : bottomLimit;
            }

            return parsedResult;
        }
    }
}